File: workers/common/font.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable

import numpy as np
from PIL import Image, ImageFont, ImageDraw
import warnings


@dataclass
class Glyph:
    char: str
    bitmap: np.ndarray  # shape: (height, width), dtype=uint8 values {0,1}
    width: int
    height: int


class BitmapFont:
    """
    Lightweight bitmap font loader for the provided sprite sheet
    at `src/assets/text/standard.bmp`.

    The sheet is a grid of glyphs with a 1px padding border and 1px
    margin between glyphs. Default glyph size is 5x7, starting at
    ASCII 32 (space), for 95 printable characters.

    Uses Pillow to avoid adding OpenCV as a dependency.
    """

    def __init__(
        self,
        image_path: Path,
        *,
        ascii_start: int = 32,
        letters: int = 95,
        padding: int = 1,
        margin: int = 1,
        letter_width: int = 5,
        letter_height: int = 7,
        foreground_is_dark: bool = True,
        auto_trim: bool = True,
        space_width: int | None = None,
    ) -> None:
        self.ascii_start = ascii_start
        self.letters = letters
        self.padding = padding
        self.margin = margin
        self.letter_width = letter_width
        self.letter_height = letter_height

        img = Image.open(image_path).convert("L")
        arr = np.array(img, dtype=np.uint8)
        # Convert to binary: foreground (ink) vs background. The provided
        # bitmap has dark foreground on light background.
        if foreground_is_dark:
            bits = (arr < 128).astype(np.uint8)
        else:
            bits = (arr >= 128).astype(np.uint8)

        self._glyphs: Dict[str, Glyph] = {}

        h, w = bits.shape
        gw, gh = letter_width, letter_height
        i = 0

        # Basic sheet validation (non-fatal): check we have enough room for rows/cols
        if w < padding + gw or h < padding + gh:
            raise ValueError(
                f"bitmap too small: {w}x{h} for glyph {gw}x{gh} with padding {padding}"
            )
        # Use stepping that matches src/utils/text.py: start at `padding` and step
        # by `letter_size + padding + margin`.
        row_step = gh + padding + margin
        col_step = gw + padding + margin
        max_y = h - gh - padding
        max_x = w - gw - padding
        for y in range(padding, max(0, max_y) + 1, row_step):
            for x in range(padding, max(0, max_x) + 1, col_step):
                if i >= letters:
                    break
                ch = chr(ascii_start + i)
                crop = bits[y : y + gh, x : x + gw].astype(np.uint8)
                if auto_trim:
                    cols = crop.any(axis=0)
                    if cols.any():
                        left = int(np.argmax(cols))
                        right = int(len(cols) - 1 - np.argmax(cols[::-1]))
                        crop = crop[:, left : right + 1]
                width = int(crop.shape[1])
                if ch == " ":
                    # ensure visible spacing for space
                    sw = space_width if space_width is not None else max(1, gw // 2)
                    crop = np.zeros((gh, sw), dtype=np.uint8)
                    width = sw
                self._glyphs[ch] = Glyph(ch, crop, width, gh)
                i += 1
            if i >= letters:
                break

        # Ensure space exists even if not present
        if " " not in self._glyphs:
            sw = space_width if space_width is not None else max(1, gw // 2)
            self._glyphs[" "] = Glyph(" ", np.zeros((gh, sw), dtype=np.uint8), sw, gh)

        # Final validation: number of glyphs
        if len(self._glyphs) < min(letters, 95):
            warnings.warn(
                f"loaded only {len(self._glyphs)} glyphs (expected around {letters}); check sheet layout and params",
                RuntimeWarning,
            )

    def get(self, ch: str) -> Glyph:
        # Fallback to space for unknown glyphs
        return self._glyphs.get(
            ch, self._glyphs.get(" ", next(iter(self._glyphs.values())))
        )

    @property
    def glyph_count(self) -> int:
        return len(self._glyphs)

    def render_text_row(self, text: str, letter_spacing: int = 1) -> np.ndarray:
        """
        Render text into a single-row bitmap of height `letter_height` and
        width equal to sum of glyph widths plus spacing.
        Returns array of shape (letter_height, total_width) with 0/1 values.
        """
        if not text:
            return np.zeros((self.letter_height, 0), dtype=np.uint8)
        parts: list[np.ndarray] = []
        first = True
        for ch in text:
            g = self.get(ch)
            if not first:
                parts.append(
                    np.zeros(
                        (self.letter_height, max(0, int(letter_spacing))),
                        dtype=np.uint8,
                    )
                )
            parts.append(g.bitmap)
            first = False
        return (
            np.concatenate(parts, axis=1)
            if parts
            else np.zeros((self.letter_height, 0), dtype=np.uint8)
        )


def repo_root_from(file: Path) -> Path:
    # workers/common/font.py -> workers/common -> workers -> repo root
    return file.resolve().parents[2]


class TTFFont:
    """
    Optional TTF-based font renderer using Pillow. Not used by default.

    Renders ASCII 32..(32+letters-1) to binary bitmaps at a target pixel height.
    Each glyph is auto-trimmed horizontally to variable width.
    """

    def __init__(
        self,
        font_path: Path,
        *,
        pixel_height: int = 10,
        ascii_start: int = 32,
        letters: int = 95,
        threshold: int = 128,
        pad_x: int = 1,
        pad_y: int = 1,
        space_width: int | None = None,
    ) -> None:
        self.ascii_start = ascii_start
        self.letters = letters
        self.letter_height = pixel_height
        self._glyphs: Dict[str, Glyph] = {}

        font = ImageFont.truetype(str(font_path), size=pixel_height)

        for i in range(letters):
            ch = chr(ascii_start + i)
            # Render onto a generous canvas
            w = pixel_height * 3
            h = pixel_height + pad_y * 2
            img = Image.new("L", (w, h), 0)
            draw = ImageDraw.Draw(img)
            draw.text((pad_x, pad_y), ch, font=font, fill=255)
            arr = np.array(img, dtype=np.uint8)
            bits = (arr >= threshold).astype(np.uint8)
            # Trim horizontally to content; keep fixed height
            cols = bits.any(axis=0)
            if cols.any():
                left = int(np.argmax(cols))
                right = int(len(cols) - 1 - np.argmax(cols[::-1]))
                crop = bits[:, left : right + 1]
            else:
                crop = np.zeros((h, 1), dtype=np.uint8)
            # Normalize to target letter height by cropping/padding vertically around content
            if crop.shape[0] != self.letter_height:
                if crop.shape[0] > self.letter_height:
                    top = max(0, (crop.shape[0] - self.letter_height) // 2)
                    crop = crop[top : top + self.letter_height, :]
                else:
                    pad_top = (self.letter_height - crop.shape[0]) // 2
                    pad_bot = self.letter_height - crop.shape[0] - pad_top
                    crop = np.pad(crop, ((pad_top, pad_bot), (0, 0)), mode="constant")
            gw = crop.shape[1]
            if ch == " ":
                sw = (
                    space_width
                    if space_width is not None
                    else max(1, pixel_height // 3)
                )
                crop = np.zeros((self.letter_height, sw), dtype=np.uint8)
                gw = sw
            self._glyphs[ch] = Glyph(ch, crop, gw, self.letter_height)

    def get(self, ch: str) -> Glyph:
        return self._glyphs.get(
            ch, self._glyphs.get(" ", next(iter(self._glyphs.values())))
        )

    def render_text_row(self, text: str, letter_spacing: int = 1) -> np.ndarray:
        if not text:
            return np.zeros((self.letter_height, 0), dtype=np.uint8)
        parts: list[np.ndarray] = []
        first = True
        for ch in text:
            g = self.get(ch)
            if not first:
                parts.append(
                    np.zeros(
                        (self.letter_height, max(0, int(letter_spacing))),
                        dtype=np.uint8,
                    )
                )
            parts.append(g.bitmap)
            first = False
        return (
            np.concatenate(parts, axis=1)
            if parts
            else np.zeros((self.letter_height, 0), dtype=np.uint8)
        )

File: workers/common/font.ts
import { dirname } from "path";
import { readFileSync } from "fs";

export type Rows = number[][]; // 0/1 ints

export type Glyph = {
  char: string;
  width: number;
  height: number;
  bitmap: Rows; // height x width, 0/1
};

export type BitmapFontOpts = {
  asciiStart?: number; // default 32
  letters?: number; // default 95
  padding?: number; // pixels around grid, default 1
  margin?: number; // pixels between glyphs, default 1
  letterWidth?: number; // default 5
  letterHeight?: number; // default 7
  foregroundIsDark?: boolean; // default true
  autoTrim?: boolean; // default true (trim empty columns to variable width)
  spaceWidth?: number | null; // default null → max(1, letterWidth/2)
};

type BMP = { width: number; height: number; bpp: number; data: Uint8Array }; // BGR triples, bottom-up

function decodeBmp(buf: Uint8Array): BMP {
  // Minimal BMP decoder for OS/2 1.x (COREHEADER) 24bpp or Windows 24bpp.
  if (buf[0] !== 0x42 || buf[1] !== 0x4d) throw new Error("Not a BMP");
  const fileSize = buf[2] | (buf[3] << 8) | (buf[4] << 16) | (buf[5] << 24);
  const dataOffset = buf[10] | (buf[11] << 8) | (buf[12] << 16) | (buf[13] << 24);
  const headerSize = buf[14] | (buf[15] << 8) | (buf[16] << 16) | (buf[17] << 24);
  let width = 0, height = 0, bpp = 0, off = 14 + headerSize;
  if (headerSize === 12) {
    // OS/2 BITMAPCOREHEADER
    width = buf[18] | (buf[19] << 8);
    height = buf[20] | (buf[21] << 8);
    bpp = buf[24] | (buf[25] << 8);
  } else {
    // Windows BITMAPINFOHEADER (or later). Read 32-bit width/height.
    width = buf[18] | (buf[19] << 8) | (buf[20] << 16) | (buf[21] << 24);
    height = buf[22] | (buf[23] << 8) | (buf[24] << 16) | (buf[25] << 24);
    bpp = buf[28] | (buf[29] << 8);
  }
  if (bpp !== 24) throw new Error(`Unsupported BMP bpp: ${bpp}`);
  const rowStride = ((width * 3 + 3) & ~3) >>> 0; // 4-byte aligned rows
  const data = new Uint8Array(width * height * 3);
  // BMP pixel data is bottom-up if height > 0
  const bottomUp = height > 0;
  const absH = Math.abs(height);
  const start = dataOffset;
  for (let y = 0; y < absH; y++) {
    const srcRow = start + y * rowStride;
    const dstY = bottomUp ? (absH - 1 - y) : y;
    for (let x = 0; x < width; x++) {
      const si = srcRow + x * 3;
      const di = (dstY * width + x) * 3;
      data[di + 0] = buf[si + 2]; // R
      data[di + 1] = buf[si + 1]; // G
      data[di + 2] = buf[si + 0]; // B
    }
  }
  return { width, height: absH, bpp, data };
}

function toBinaryRows(bmp: BMP, foregroundIsDark: boolean): Rows {
  const { width, height, data } = bmp;
  const rows: Rows = Array.from({ length: height }, () => Array(width).fill(0));
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * 3;
      const r = data[i + 0], g = data[i + 1], b = data[i + 2];
      const gray = Math.round(0.299 * r + 0.587 * g + 0.114 * b);
      rows[y][x] = foregroundIsDark ? (gray < 128 ? 1 : 0) : (gray >= 128 ? 1 : 0);
    }
  }
  return rows;
}

export class BitmapFont {
  private glyphs = new Map<string, Glyph>();
  readonly letterHeight: number;
  readonly letterWidth: number;
  constructor(
    imagePath: string,
    opts: BitmapFontOpts = {}
  ) {
    const asciiStart = opts.asciiStart ?? 32;
    const letters = opts.letters ?? 95;
    const padding = opts.padding ?? 1;
    const margin = opts.margin ?? 1;
    const letterWidth = opts.letterWidth ?? 5;
    const letterHeight = opts.letterHeight ?? 7;
    const foregroundIsDark = opts.foregroundIsDark ?? true;
    const autoTrim = opts.autoTrim ?? true;
    const spaceWidth = opts.spaceWidth ?? null;

    this.letterHeight = letterHeight;
    this.letterWidth = letterWidth;

    let buf: Uint8Array;
    try {
      const raw = readFileSync(imagePath);
      buf = raw instanceof Uint8Array ? raw : new Uint8Array(raw.buffer);
    } catch (e) {
      throw new Error(`Font BMP not found: ${imagePath}`);
    }
    const bmp = decodeBmp(buf);
    const bits = toBinaryRows(bmp, foregroundIsDark);

    const h = bits.length;
    const w = bits[0]?.length ?? 0;
    const gw = letterWidth, gh = letterHeight;
    if (w < padding + gw || h < padding + gh) throw new Error(`bitmap too small: ${w}x${h} for glyph ${gw}x${gh}`);

    const rowStep = gh + padding + margin;
    const colStep = gw + padding + margin;
    const maxY = h - gh - padding;
    const maxX = w - gw - padding;
    let i = 0;
    for (let y = padding; y <= Math.max(0, maxY) && i < letters; y += rowStep) {
      for (let x = padding; x <= Math.max(0, maxX) && i < letters; x += colStep) {
        const ch = String.fromCharCode(asciiStart + i);
        // crop gw x gh
        const crop: Rows = Array.from({ length: gh }, (_, yy) => bits[y + yy].slice(x, x + gw));
        let trimmed = crop;
        if (autoTrim) {
          const cols = Array.from({ length: gw }, (_, cx) => crop.some(row => row[cx] === 1));
          if (cols.some(Boolean)) {
            let left = 0; while (left < cols.length && !cols[left]) left++;
            let right = cols.length - 1; while (right >= 0 && !cols[right]) right--;
            trimmed = crop.map(row => row.slice(left, right + 1));
          }
        }
        let width = trimmed[0]?.length ?? gw;
        let bitmap = trimmed;
        if (ch === ' ') {
          const sw = (spaceWidth !== null && spaceWidth !== undefined) ? Number(spaceWidth) : Math.max(1, Math.floor(gw / 2));
          width = sw;
          bitmap = Array.from({ length: gh }, () => Array(sw).fill(0));
        }
        this.glyphs.set(ch, { char: ch, bitmap, width, height: gh });
        i++;
      }
    }
    if (!this.glyphs.has(' ')) {
      const sw = (spaceWidth !== null && spaceWidth !== undefined) ? Number(spaceWidth) : Math.max(1, Math.floor(gw / 2));
      this.glyphs.set(' ', { char: ' ', bitmap: Array.from({ length: gh }, () => Array(sw).fill(0)), width: sw, height: gh });
    }
  }

  get(ch: string): Glyph {
    return this.glyphs.get(ch) || this.glyphs.get(' ') || { char: ' ', width: 1, height: this.letterHeight, bitmap: Array.from({ length: this.letterHeight }, () => [0]) };
  }

  get glyphCount(): number { return this.glyphs.size; }

  renderTextRow(text: string, letterSpacing = 1): Rows {
    if (!text) return Array.from({ length: this.letterHeight }, () => [] as number[]);
    const parts: Rows[] = [];
    let first = true;
    for (const ch of text) {
      const g = this.get(ch);
      if (!first) parts.push(Array.from({ length: this.letterHeight }, () => Array(Math.max(0, letterSpacing)).fill(0)));
      parts.push(g.bitmap);
      first = false;
    }
    // Concatenate horizontally
    const totalW = parts.reduce((acc, r) => acc + (r[0]?.length ?? 0), 0);
    const out: Rows = Array.from({ length: this.letterHeight }, () => Array(totalW).fill(0));
    let x0 = 0;
    for (const block of parts) {
      const w = block[0]?.length ?? 0;
      for (let y = 0; y < this.letterHeight; y++) {
        for (let x = 0; x < w; x++) out[y][x0 + x] = block[y][x];
      }
      x0 += w;
    }
    return out;
  }
}

export function repoRootFromWorkers(): string {
  // When spawned with cwd 'workers', repo root is parent
  return dirname(process.cwd());
}

export function resolveFontPathFromRoot(root: string): string {
  const primary = `${root}/assets/text/standard.bmp`;
  // Throw if missing to surface misconfiguration clearly
  readFileSync(primary);
  return primary;
}
